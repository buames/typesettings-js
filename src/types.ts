// tslint:disable max-line-length
import {
  FontFamilyProperty,
  FontSizeProperty,
  FontStyleProperty,
  FontWeightProperty,
  LineHeightProperty,
  LetterSpacingProperty,
  TextTransformProperty
} from 'csstype';

export enum FontSourceFormats {
  locals = 'locals',
  woff2 = 'woff2',
  woff = 'woff',
  ttf = 'ttf',
  otf = 'opentype',
  eot = 'embedded-opentype'
}

export enum LetterCasing {
  normalcase = 'normalcase',
  uppercase = 'uppercase',
  lowercase = 'lowercase'
}

export type StyledValue = string | number | undefined;

export type StyledObject = { [property: string]: StyledValue };

export type FontSources = { [K in FontSourceFormats]?: string | string[] | NodeRequireFunction | false };

export type TypesettingResults = { [size: string]: { [weight: string]: StyledFont } };

export interface TypesettingOptions {
  [k: string]: any;
  cssFn?: (styles: StyledObject | string) => any;
  fontStyles?: StyledObject;
  fontFaceStyles?: StyledObject;
}

export interface Typesettings {
  family: FontFamilyProperty;
  variants: FontVariant[];
  fallbacks?: FontFamilyProperty[];
}

export interface FontVariant {
  fontStyle: FontStyleProperty;
  fontWeight: FontWeightProperty;
  sources: FontSources;
  normalcase?: FontSetting[];
  uppercase?: FontSetting[];
  lowercase?: FontSetting[];
}

export interface FontSetting {
  fontSize: FontSizeProperty<StyledValue>;
  letterSpacing?: LetterSpacingProperty<StyledValue> | null;
  lineHeight?: LineHeightProperty<StyledValue> | null;
}

export interface StyledFont extends StyledObject {
  fontFamily: FontFamilyProperty;
  fontStyle?: FontStyleProperty;
  fontSize?: FontSizeProperty<StyledValue>;
  fontWeight?: FontWeightProperty;
  letterSpacing?: LetterSpacingProperty<StyledValue>;
  lineHeight?: LineHeightProperty<StyledValue>;
  textTransform?: TextTransformProperty;
}
