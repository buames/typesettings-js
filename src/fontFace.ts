import { Typesettings, FontSourceFormats, TypesettingOptions } from './types';
import { getFamilyName } from './utils';

export const generateFontFace = (
  typesettings: Typesettings,
  options: TypesettingOptions = { }
): string => {
  const { family, variants } = typesettings;
  const declaration = variants.map(({ sources, fontStyle, fontWeight }) => {
    if (!sources || (sources && Object.keys(sources).length === 0)) {
      throw Error('Missing font file sources');
    }

    const srcs = Object.keys(sources).map(key => (
      sources[key] && (Array.isArray(sources[key])
        ? sources[key].map((n: string) => (`local('${ n }')`))
        : `url(${ sources[key] }) format('${ FontSourceFormats[key] }')`)
    )).filter(Boolean);

    const face = [
      `font-family: ${ getFamilyName(family) }`,
      fontStyle && `font-style: ${ fontStyle }`,
      fontWeight && `font-weight: ${ fontWeight }`,
      srcs && `src: ${ srcs.join(', ') };`,
      options.fontFaceStyles
    ].filter(Boolean);

    return `@font-face { ${ face.join('; ') } }`;
  });

  const fontFace = declaration.join(' ');
  return options.cssFn ? options.cssFn(fontFace) : fontFace;
};
