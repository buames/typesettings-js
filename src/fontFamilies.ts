import { FontFamilyProperty } from 'csstype';

const genericFamilyKeywords = [
  'initial',
  'inherit',
  'unset',
  'serif',
  'sans-serif',
  'cursive',
  'fantasy',
  'monospace',
  'system-ui',
];

/**
 * Checks whether or not a font family name is a generic font family keyword
 *
 * When a generic family keyword keyword is wrapped in quotes, the browser will not interpret
 * them as keywords. Instead it looks for a font by that name (e.g. "sans-serif" font),
 * which is likley not what you actually want.
 *
 * @see https://mathiasbynens.be/notes/unquoted-font-family
 */
const isKeyword = (family: FontFamilyProperty) => {
  return genericFamilyKeywords.includes(family.toLowerCase());
};

/**
 * Checks whether or not a font family name is a vendor prefixed system font
 *
 * Quotes should never be used around vendor prefixed system fonts since the browser
 * will not interpret them as keywords
 */
const isSystemFont = (family: FontFamilyProperty) => {
  return family.startsWith('-apple-') || family === 'BlinkMacSystemFont';
};

/**
 * Checks whether or not a font family name should be wrapped in quotes
 *
 * It is recommended to quote font family names that contain white space, digits, or
 * punctuation characters other than hyphens.
 *
 * @see https://www.w3.org/TR/CSS2/fonts.html#font-family-prop
 */
const isCharactersOnly = (family: FontFamilyProperty) => {
  return /^[-a-zA-Z]+$/.test(family);
};

/**
 * Normalizes quotations for a given font family name or string of font family names.
 */
export const normalizeFontFamily = (value: FontFamilyProperty) => {
  return value
    .split(',')
    .map((raw: FontFamilyProperty) => {
      // Clean the family name, includes its quotes
      const family = raw.trim().replace(/^['"]|['"]$/g, '');

      if (
        isKeyword(family) ||
        isSystemFont(family) ||
        isCharactersOnly(family)
      ) {
        return family;
      }

      return `'${family}'`;
    }, [])
    .join(', ');
};

/**
 * Returns the font stack from a given family and array of fallbacks
 * */
export const getFontStack = (
  family: FontFamilyProperty,
  fallbacks?: FontFamilyProperty[],
) => {
  const stack = [
    normalizeFontFamily(family),
    fallbacks?.map(normalizeFontFamily).join(', '),
  ];

  return stack.filter(Boolean).join(', ');
};
