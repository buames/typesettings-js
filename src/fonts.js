/* eslint-disable object-curly-newline, no-param-reassign */
const merge = require('deepmerge')

/*
  Returns a weight property labelâ€š prefixed with 'n' for normal or 'i' for italics.
  For example, n700 equals 'normal' with a weight of '700'.
*/
const getStyleLabel = (fontStyle, fontWeight) => {
  switch (fontStyle) {
    case 'normal': return `n${ fontWeight }`
    case 'italics': return `i${ fontWeight }`
    default: return ''
  }
}

/*
  Returns a property label to append to the style property
  label depending on the lettercasing type
*/
const getTransformLabel = (lettercasing) => {
  switch (lettercasing) {
    case 'uppercase': return '_caps'
    case 'lowercase': return '_lower'
    case 'normalcase': return ''
    default: return ''
  }
}

/*
  Generates a map of typesettings. We do not return emotionjs classes
  as it would not work with media queries properly.
*/
const generate = (variant, options) => {
  const { casing, styles, family, fallback } = options
  const sizes = variant[casing]

  if (!sizes) {
    return { }
  }

  return Object.keys(sizes).reduce((accum, size) => {
    const { characterSpacing, lineHeight } = sizes[size]
    const styleLabel = getStyleLabel(variant.fontStyle, variant.fontWeight)
    const transformLabel = getTransformLabel(casing)

    accum[`s${ size }`] = { } // font-size property is prefixed with 's'
    accum[`s${ size }`][`${ styleLabel }${ transformLabel }`] = {
      fontFamily: fallback ? `${ family }, ${ fallback }` : family,
      fontSize: `${ size }px`,
      fontStyle: variant.fontStyle,
      fontWeight: variant.fontWeight,
      letterSpacing: `${ characterSpacing ? `${ characterSpacing.toFixed(2) }px` : 'initial' }`,
      lineHeight: `${ lineHeight ? `${ lineHeight }px` : 'normal' }`,
      textTransform: `${ casing === 'normalcase' ? 'none' : casing }`,
      ...styles || { }
    }

    return accum
  }, { })
}

const generateFonts = (typesettings, styles) => {
  if (typeof typesettings !== 'object') {
    throw new Error('Your typesettings must be an object.')
  }

  const { family, fallback, ...variants } = typesettings

  // Loop over the typeface variants and create each variant's
  // typesettings for each of their lettingcasing
  const settings = Object.values(variants).map((variant) => {
    const shared = { styles, family, fallback }
    const normal = generate(variant, { casing: 'normalcase', ...shared })
    const upper = generate(variant, { casing: 'uppercase', ...shared })
    const lower = generate(variant, { casing: 'lowercase', ...shared })
    return merge.all([ normal, upper, lower ])
  })

  return merge.all(settings)
}

export default generateFonts
