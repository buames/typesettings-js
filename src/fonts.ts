import {
  LetterCasing,
  StyledValue,
  Typesettings,
  TypesettingFontsOptions,
  TypesettingResults,
  FontVariant,
  FontSetting
} from './types';

import {
  getFontStack,
  parseSize
} from './utils';

/*
  Returns a pixel value or the raw css value
*/
const px = (n: StyledValue) => (
  typeof n === 'number' && n !== 0 ? `${ n }px` : n
);

/*
  Returns a weight property labelâ€š prefixed with 'n' for normal, 'i' for italics,
  and 'o' for oblique. For example, n700 equals 'normal' with a weight of '700'.
  However, if the weight is a string (ie 'bold'), this returns the prefix + a capitalized
  weight. For example, nBold.
*/
const getStyleLabel = ({ fontStyle, fontWeight }: FontVariant) => (
  `${ fontStyle.charAt(0) }${ typeof fontWeight === 'string'
    ? `${ fontWeight.charAt(0).toUpperCase() }${ fontWeight.slice(1) }`
    : fontWeight }`
);

/*
  Returns a property label to append to the style property
  label depending on the lettercasing type
*/
const getTransformLabel = (casing: LetterCasing | string) => (
  (casing === LetterCasing.uppercase) ? '_caps'
    : (casing === LetterCasing.lowercase) ? '_lower'
    : ''
);

/*
  Generates a map of typesettings
*/
export const generateFonts = (
  typesettings: Typesettings,
  options: TypesettingFontsOptions = { }
): TypesettingResults => {
  const { family, fallbacks, variants } = typesettings;
  const fontFamily = getFontStack(family, fallbacks);

  const styles: { } = variants.reduce((acc, variant) => {
    const { fontStyle, fontWeight, sources, ...casings } = variant;
    const styleLabel = getStyleLabel(variant);

    // Loop over each casing (ie. normalcase, uppercase, lowercase)
    Object.keys(casings).forEach((casing) => {
      const transformLabel = getTransformLabel(casing);

      // Now loop over each style object
      variant[casing].forEach((setting: FontSetting) => {
        const sizeLabel = `s${ parseSize(setting.fontSize) }`;
        const weightLabel = `${ styleLabel }${ transformLabel }`;

        acc[sizeLabel] = acc[sizeLabel] || { };
        acc[sizeLabel][weightLabel] = {
          fontFamily,
          fontStyle,
          fontWeight,
          fontSize: px(setting.fontSize),
          letterSpacing: setting.letterSpacing && px(setting.letterSpacing),
          lineHeight: setting.lineHeight && px(setting.lineHeight),
          textTransform: casing !== LetterCasing.normalcase ? casing : 'none',
          ...options.styles || { }
        };

        if (options.cssFn) {
          acc[sizeLabel][weightLabel] = options.cssFn(acc[sizeLabel][weightLabel]);
        }
      });
    });

    return acc;
  // tslint:disable-next-line: align
  }, { });

  return styles;
};
