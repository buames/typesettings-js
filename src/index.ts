/* eslint-disable no-nested-ternary */
import { FontFamilyProperty, FontSizeProperty } from 'csstype';
import {
  fontCasings,
  fontSources,
  FontFaceOptions,
  FontStyleOptions,
  FontSources,
  FontSetting,
  FontVariant,
  StyledCssFn,
  StyledValue,
  StyledObject,
  Typesettings,
  TypesettingOptions,
  TypesettingsFontsResult,
  TypesettingsFontFaceResult,
} from './types';

export {
  fontCasings,
  fontSources,
  FontFaceOptions,
  FontStyleOptions,
  FontSources,
  FontSetting,
  FontVariant,
  StyledCssFn,
  StyledValue,
  StyledObject,
  Typesettings,
  TypesettingOptions,
  TypesettingsFontsResult,
  TypesettingsFontFaceResult,
};

/**
 * Converts a given value to a `pixel` unit
 * */
export const px = (n: StyledValue) =>
  typeof n === 'number' && n !== 0 ? `${n}px` : n;

/**
 * Parses a number and unit string, returning only the number used
 * */
export const parseSize = (str: FontSizeProperty<StyledValue>) => {
  if (typeof str !== 'string') return str;

  const match = str.trim().match(/([\d.\-+]*)\s*(.*)/);
  return (match && match[1]) || str;
};

/**
 * Returns a normalized FontFamily name where names with
 * a space are automatically wrapped in quotes
 * */
export const normalizeFamily = (family: FontFamilyProperty) =>
  /\s/g.test(family) ? `'${family}'` : family;

/**
 * Returns the font stack from a given family and array of fallbacks
 * */
export const getFontStack = (
  family: FontFamilyProperty,
  fallbacks?: FontFamilyProperty[],
) =>
  [
    normalizeFamily(family),
    fallbacks && fallbacks.map(normalizeFamily).join(', '),
  ]
    .filter(Boolean)
    .join(', ');

/**
 * Generate font styles from a typesettings obj
 * */
export const generateFonts = <T>(
  typesettings: Typesettings,
  options: TypesettingOptions<T> = {},
): TypesettingsFontsResult<T> => {
  /*
    Returns a weight property labelâ€š prefixed with 'n' for normal, 'i' for italics,
    and 'o' for oblique. For example, n700 equals 'normal' with a weight of '700'.
    However, if the weight is a string (ie 'bold'), this returns the prefix + a capitalized
    weight. For example, nBold.
  */
  const getStyleLabel = ({ fontStyle, fontWeight }: FontVariant) =>
    `${fontStyle.charAt(0)}${
      typeof fontWeight === 'string'
        ? `${fontWeight.charAt(0).toUpperCase()}${fontWeight.slice(1)}`
        : fontWeight
    }`;

  /*
  Returns a property label to append to the style property
  label depending on the lettercasing type
  */
  const getTransformLabel = (casing: typeof fontCasings | string) =>
    casing === fontCasings.uppercase
      ? '_caps'
      : casing === fontCasings.lowercase
      ? '_lower'
      : '';

  const { family, fallbacks, variants } = typesettings;
  const fontFamily = getFontStack(family, fallbacks);

  const styles = variants.reduce((acc, variant) => {
    const { fontStyle, fontWeight, sources, ...casings } = variant;

    const styleLabel = getStyleLabel(variant);

    // Loop over each casing (ie. normalcase, uppercase, lowercase)
    (Object.keys(casings) as (keyof typeof casings)[]).forEach((casing) => {
      const transformLabel = getTransformLabel(casing);
      const values = variant[casing];

      if (values) {
        values.forEach((setting: FontSetting) => {
          const sizeLabel = `s${parseSize(setting.fontSize)}`;
          const weightLabel = `${styleLabel}${transformLabel}`;

          acc[sizeLabel] = acc[sizeLabel] || {};
          acc[sizeLabel][weightLabel] = {
            fontFamily,
            fontStyle,
            fontWeight,
            fontSize: px(setting.fontSize),
            letterSpacing: setting.letterSpacing
              ? px(setting.letterSpacing)
              : 'initial',
            lineHeight: setting.lineHeight ? px(setting.lineHeight) : 'normal',
            ...(casing !== fontCasings.normalcase && {
              textTransform: casing,
            }),
            ...(options.fontStyles || {}),
          };

          if (typeof options.cssFn === 'function') {
            acc[sizeLabel][weightLabel] = options.cssFn(
              acc[sizeLabel][weightLabel],
            );
          }
        });
      }
    });

    return acc;
  }, {} as TypesettingsFontsResult<StyledObject | StyledCssFn>);

  return (styles as unknown) as TypesettingsFontsResult<T>;
};

/**
 * Generate a @font-face declaration from a typesettings obj
 * */
export const generateFontFace = <T>(
  typesettings: Typesettings,
  options: TypesettingOptions<T> = {},
): TypesettingsFontFaceResult<T> => {
  const { family, variants } = typesettings;

  const declaration = variants.map(({ sources, fontStyle, fontWeight }) => {
    const srcs = (Object.keys(sources) as (keyof typeof sources)[])
      .map((key) => {
        const src = sources[key];
        const format = fontSources[key];

        if (!src) return;

        if (Array.isArray(src) && key === 'locals') {
          return src.map((n: string) => `local('${n}')`);
        }

        if (key === 'eot') {
          return `url(${src}?#iefix) format('${format}')`;
        }

        return `url(${src}) format('${format}')`;
      })
      .filter(Boolean);

    const styles = {
      fontStyle,
      fontWeight,
      fontFamily: normalizeFamily(family),
      src: srcs.join(', '),
      ...options.fontFaceStyles,
    };

    const face = Object.entries(styles).reduce((acc, [key, value]) => {
      const propName = key.replace(
        /([A-Z])/g,
        (match) => `-${match[0].toLowerCase()}`,
      );
      return `${acc}${propName}: ${value};`;
    }, '');

    return `@font-face { ${face} }`;
  });

  const fontFace = declaration.join(' ');

  return typeof options.cssFn === 'function'
    ? options.cssFn(fontFace)
    : fontFace;
};

/**
 * Convenience function to generate font styles and a @font-face declaration
 * */
export const generate = <T>(
  typesettings: Typesettings,
  options?: TypesettingOptions<T>,
) => ({
  fontFace: generateFontFace<T>(typesettings, options),
  fonts: generateFonts<T>(typesettings, options),
});
