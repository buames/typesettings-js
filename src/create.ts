import { Typesettings, TypesettingOptions } from './types';
import { createFontFace } from './fontFace';
import { createFonts } from './fonts';

/**
 * Convenience function to create font styles and a @font-face declaration
 * */
export const create = <T>(
  typesettings: Typesettings,
  options?: TypesettingOptions<T>,
) => ({
  fontFace: createFontFace<T>(typesettings, options),
  fonts: createFonts<T>(typesettings, options),
});
