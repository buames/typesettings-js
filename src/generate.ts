import { Typesettings, TypesettingOptions, TypesettingResults } from './types';
import { generateFontFace } from './fontFace';
import { generateFonts } from './fonts';

export const generate = (
  typesettings: Typesettings,
  options?: TypesettingOptions
): { fontFace: string, fonts: TypesettingResults } => ({
  fontFace: generateFontFace(typesettings, options),
  fonts: generateFonts(typesettings, options)
});
