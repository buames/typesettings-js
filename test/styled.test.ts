import styled from '@emotion/styled';
import { css, Interpolation, SerializedStyles } from '@emotion/core';
import { build } from './fixtures/a';
import { createFontFace, createFonts } from '../src';

type StyledCssFn = (...args: Interpolation[]) => SerializedStyles;

test('Should return @font-face declaration as the return type of a cssFn', () => {
  const a = createFontFace<StyledCssFn>(build(), { cssFn: css });
  const b = createFontFace(build());
  expect(a.styles).toEqual(b);
});

test('Should return @font-face declaration snapshots', () => {
  const a = createFontFace(build());
  expect(a).toMatchSnapshot('@font-face declaration when cssFn is undefined');

  const b = createFontFace<StyledCssFn>(build(), { cssFn: css });
  expect(b).toMatchSnapshot('@font-face declaration when cssFn is set');
});

test('Should return font style as the return type of a cssFn', () => {
  const a = createFonts<StyledCssFn>(build({ fallbacks: undefined }), {
    cssFn: css,
  });
  expect(a.s14.i400.styles).toBe(
    'font-family:Helvetica;font-style:italic;font-weight:400;font-size:14px;letter-spacing:initial;line-height:18px;',
  );
});

test('Should return font style snapshots', () => {
  const a = createFonts(build());
  expect(a).toMatchSnapshot(
    'Font styles default return type when cssFn is undefined',
  );

  const a1 = styled('p')`
    ${a.s14.n400} color: tomato;
  `;
  expect(a1).toMatchSnapshot(
    '`styled` Interpolation font styles when cssFn is undefined',
  );

  const a2 = styled('p')(a.s14.n400, { color: 'tomato' });
  expect(a2).toMatchSnapshot(
    '`styled` ObjectInterpolation font styles when cssFn is undefined',
  );

  const a3 = styled('p')({ ...a.s14.n400, color: 'tomato' });
  expect(a3).toMatchSnapshot(
    '`styled` ObjectInterpolation font styles (spread) when cssFn is undefined',
  );

  const a4 = styled('p')([a.s14.n400, { color: 'tomato' }]);
  expect(a4).toMatchSnapshot(
    '`styled` ArrayInterpolation font styles when cssFn is undefined',
  );

  const b = createFonts<StyledCssFn>(build(), { cssFn: css });
  expect(b).toMatchSnapshot(
    'Font styles default return type when cssFn is set',
  );

  const b1 = styled('p')`
    ${b.s14.n400} color: tomato;
  `;
  expect(b1).toMatchSnapshot(
    '`styled` Interpolation font styles when cssFn is set',
  );

  const b2 = styled('p')(b.s14.n400, { color: 'tomato' });
  expect(b2).toMatchSnapshot(
    '`styled` ObjectInterpolation font styles when cssFn is set',
  );

  const b3 = styled('p')({ ...b.s14.n400, color: 'tomato' });
  expect(b3).toMatchSnapshot(
    '`styled` ObjectInterpolation font styles (spread) when cssFn is set',
  );

  const b4 = styled('p')([b.s14.n400, { color: 'tomato' }]);
  expect(b4).toMatchSnapshot(
    '`styled` ArrayInterpolation font styles when cssFn is set',
  );
});
