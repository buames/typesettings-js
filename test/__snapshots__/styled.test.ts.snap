// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Should return @font-face declaration snapshots: @font-face declaration when cssFn is set 1`] = `
Object {
  "map": undefined,
  "name": "jj2j4f",
  "next": undefined,
  "styles": "@font-face { font-style: italic;font-weight: 400;font-family: Helvetica;src: local('Helvetica Regular'),local('Helvetica-Regular'); } @font-face { font-style: normal;font-weight: 400;font-family: Helvetica;src: local('Helvetica Regular'),local('Helvetica-Regular'); } @font-face { font-style: normal;font-weight: 500;font-family: Helvetica;src: local('Helvetica Medium'),local('Helvetica-Medium'), url(./font-file.eot?#iefix) format('embedded-opentype'), url(./font-file.woff) format('woff'), url(./font-file.woff2) format('woff2'), url(./font-file.ttf) format('ttf'); } @font-face { font-style: normal;font-weight: bold;font-family: Helvetica;src: local('Helvetica Bold'),local('Helvetica-Bold'), url(./font-file.eot?#iefix) format('embedded-opentype'), url(./font-file.woff) format('woff'), url(./font-file.woff2) format('woff2'), url(./font-file.ttf) format('ttf'); }",
  "toString": [Function],
}
`;

exports[`Should return @font-face declaration snapshots: @font-face declaration when cssFn is undefined 1`] = `"@font-face { font-style: italic;font-weight: 400;font-family: Helvetica;src: local('Helvetica Regular'),local('Helvetica-Regular'); } @font-face { font-style: normal;font-weight: 400;font-family: Helvetica;src: local('Helvetica Regular'),local('Helvetica-Regular'); } @font-face { font-style: normal;font-weight: 500;font-family: Helvetica;src: local('Helvetica Medium'),local('Helvetica-Medium'), url(./font-file.eot?#iefix) format('embedded-opentype'), url(./font-file.woff) format('woff'), url(./font-file.woff2) format('woff2'), url(./font-file.ttf) format('ttf'); } @font-face { font-style: normal;font-weight: bold;font-family: Helvetica;src: local('Helvetica Bold'),local('Helvetica-Bold'), url(./font-file.eot?#iefix) format('embedded-opentype'), url(./font-file.woff) format('woff'), url(./font-file.woff2) format('woff2'), url(./font-file.ttf) format('ttf'); }"`;

exports[`Should return font style snapshots: \`styled\` ArrayInterpolation font styles when cssFn is set 1`] = `
Object {
  "$$typeof": Symbol(react.forward_ref),
  "__emotion_base": "p",
  "__emotion_forwardProp": undefined,
  "__emotion_real": [Circular],
  "__emotion_styles": Array [
    Array [
      Object {
        "map": undefined,
        "name": "1ezn3fw",
        "next": undefined,
        "styles": "font-family:Helvetica, -apple-system, BlinkMacSystemFont;font-style:normal;font-weight:400;font-size:14px;letter-spacing:initial;line-height:18px;",
        "toString": [Function],
      },
      Object {
        "color": "tomato",
      },
    ],
  ],
  "defaultProps": undefined,
  "displayName": "Styled(p)",
  "render": [Function],
  "withComponent": [Function],
}
`;

exports[`Should return font style snapshots: \`styled\` ArrayInterpolation font styles when cssFn is undefined 1`] = `
Object {
  "$$typeof": Symbol(react.forward_ref),
  "__emotion_base": "p",
  "__emotion_forwardProp": undefined,
  "__emotion_real": [Circular],
  "__emotion_styles": Array [
    Array [
      Object {
        "fontFamily": "Helvetica, -apple-system, BlinkMacSystemFont",
        "fontSize": "14px",
        "fontStyle": "normal",
        "fontWeight": 400,
        "letterSpacing": "initial",
        "lineHeight": "18px",
      },
      Object {
        "color": "tomato",
      },
    ],
  ],
  "defaultProps": undefined,
  "displayName": "Styled(p)",
  "render": [Function],
  "withComponent": [Function],
}
`;

exports[`Should return font style snapshots: \`styled\` Interpolation font styles when cssFn is set 1`] = `
Object {
  "$$typeof": Symbol(react.forward_ref),
  "__emotion_base": "p",
  "__emotion_forwardProp": undefined,
  "__emotion_real": [Circular],
  "__emotion_styles": Array [
    "
    ",
    Object {
      "map": undefined,
      "name": "1ezn3fw",
      "next": undefined,
      "styles": "font-family:Helvetica, -apple-system, BlinkMacSystemFont;font-style:normal;font-weight:400;font-size:14px;letter-spacing:initial;line-height:18px;",
      "toString": [Function],
    },
    " color: tomato;
  ",
  ],
  "defaultProps": undefined,
  "displayName": "Styled(p)",
  "render": [Function],
  "withComponent": [Function],
}
`;

exports[`Should return font style snapshots: \`styled\` Interpolation font styles when cssFn is undefined 1`] = `
Object {
  "$$typeof": Symbol(react.forward_ref),
  "__emotion_base": "p",
  "__emotion_forwardProp": undefined,
  "__emotion_real": [Circular],
  "__emotion_styles": Array [
    "
    ",
    Object {
      "fontFamily": "Helvetica, -apple-system, BlinkMacSystemFont",
      "fontSize": "14px",
      "fontStyle": "normal",
      "fontWeight": 400,
      "letterSpacing": "initial",
      "lineHeight": "18px",
    },
    " color: tomato;
  ",
  ],
  "defaultProps": undefined,
  "displayName": "Styled(p)",
  "render": [Function],
  "withComponent": [Function],
}
`;

exports[`Should return font style snapshots: \`styled\` ObjectInterpolation font styles (spread) when cssFn is set 1`] = `
Object {
  "$$typeof": Symbol(react.forward_ref),
  "__emotion_base": "p",
  "__emotion_forwardProp": undefined,
  "__emotion_real": [Circular],
  "__emotion_styles": Array [
    Object {
      "color": "tomato",
      "map": undefined,
      "name": "1ezn3fw",
      "next": undefined,
      "styles": "font-family:Helvetica, -apple-system, BlinkMacSystemFont;font-style:normal;font-weight:400;font-size:14px;letter-spacing:initial;line-height:18px;",
      "toString": [Function],
    },
  ],
  "defaultProps": undefined,
  "displayName": "Styled(p)",
  "render": [Function],
  "withComponent": [Function],
}
`;

exports[`Should return font style snapshots: \`styled\` ObjectInterpolation font styles (spread) when cssFn is undefined 1`] = `
Object {
  "$$typeof": Symbol(react.forward_ref),
  "__emotion_base": "p",
  "__emotion_forwardProp": undefined,
  "__emotion_real": [Circular],
  "__emotion_styles": Array [
    Object {
      "color": "tomato",
      "fontFamily": "Helvetica, -apple-system, BlinkMacSystemFont",
      "fontSize": "14px",
      "fontStyle": "normal",
      "fontWeight": 400,
      "letterSpacing": "initial",
      "lineHeight": "18px",
    },
  ],
  "defaultProps": undefined,
  "displayName": "Styled(p)",
  "render": [Function],
  "withComponent": [Function],
}
`;

exports[`Should return font style snapshots: \`styled\` ObjectInterpolation font styles when cssFn is set 1`] = `
Object {
  "$$typeof": Symbol(react.forward_ref),
  "__emotion_base": "p",
  "__emotion_forwardProp": undefined,
  "__emotion_real": [Circular],
  "__emotion_styles": Array [
    Object {
      "map": undefined,
      "name": "1ezn3fw",
      "next": undefined,
      "styles": "font-family:Helvetica, -apple-system, BlinkMacSystemFont;font-style:normal;font-weight:400;font-size:14px;letter-spacing:initial;line-height:18px;",
      "toString": [Function],
    },
    Object {
      "color": "tomato",
    },
  ],
  "defaultProps": undefined,
  "displayName": "Styled(p)",
  "render": [Function],
  "withComponent": [Function],
}
`;

exports[`Should return font style snapshots: \`styled\` ObjectInterpolation font styles when cssFn is undefined 1`] = `
Object {
  "$$typeof": Symbol(react.forward_ref),
  "__emotion_base": "p",
  "__emotion_forwardProp": undefined,
  "__emotion_real": [Circular],
  "__emotion_styles": Array [
    Object {
      "fontFamily": "Helvetica, -apple-system, BlinkMacSystemFont",
      "fontSize": "14px",
      "fontStyle": "normal",
      "fontWeight": 400,
      "letterSpacing": "initial",
      "lineHeight": "18px",
    },
    Object {
      "color": "tomato",
    },
  ],
  "defaultProps": undefined,
  "displayName": "Styled(p)",
  "render": [Function],
  "withComponent": [Function],
}
`;

exports[`Should return font style snapshots: Font styles default return type when cssFn is set 1`] = `
Object {
  "s14": Object {
    "i400": Object {
      "map": undefined,
      "name": "lhonde",
      "next": undefined,
      "styles": "font-family:Helvetica, -apple-system, BlinkMacSystemFont;font-style:italic;font-weight:400;font-size:14px;letter-spacing:initial;line-height:18px;",
      "toString": [Function],
    },
    "n400": Object {
      "map": undefined,
      "name": "1ezn3fw",
      "next": undefined,
      "styles": "font-family:Helvetica, -apple-system, BlinkMacSystemFont;font-style:normal;font-weight:400;font-size:14px;letter-spacing:initial;line-height:18px;",
      "toString": [Function],
    },
  },
  "s20": Object {
    "n500": Object {
      "map": undefined,
      "name": "y7zql4",
      "next": undefined,
      "styles": "font-family:Helvetica, -apple-system, BlinkMacSystemFont;font-style:normal;font-weight:500;font-size:20px;letter-spacing:0.29px;line-height:22px;",
      "toString": [Function],
    },
    "n500_caps": Object {
      "map": undefined,
      "name": "dmcrvq",
      "next": undefined,
      "styles": "font-family:Helvetica, -apple-system, BlinkMacSystemFont;font-style:normal;font-weight:500;font-size:20px;letter-spacing:1px;line-height:1px;text-transform:uppercase;",
      "toString": [Function],
    },
    "n500_lower": Object {
      "map": undefined,
      "name": "1yn3z8j",
      "next": undefined,
      "styles": "font-family:Helvetica, -apple-system, BlinkMacSystemFont;font-style:normal;font-weight:500;font-size:20px;letter-spacing:initial;line-height:normal;text-transform:lowercase;",
      "toString": [Function],
    },
    "nBold": Object {
      "map": undefined,
      "name": "qgokl4",
      "next": undefined,
      "styles": "font-family:Helvetica, -apple-system, BlinkMacSystemFont;font-style:normal;font-weight:bold;font-size:20rem;letter-spacing:0.29em;line-height:22em;",
      "toString": [Function],
    },
  },
  "sFooBar": Object {
    "nBold": Object {
      "map": undefined,
      "name": "jc2bz6",
      "next": undefined,
      "styles": "font-family:Helvetica, -apple-system, BlinkMacSystemFont;font-style:normal;font-weight:bold;font-size:FooBar;letter-spacing:0.29em;line-height:22em;",
      "toString": [Function],
    },
  },
}
`;

exports[`Should return font style snapshots: Font styles default return type when cssFn is undefined 1`] = `
Object {
  "s14": Object {
    "i400": Object {
      "fontFamily": "Helvetica, -apple-system, BlinkMacSystemFont",
      "fontSize": "14px",
      "fontStyle": "italic",
      "fontWeight": 400,
      "letterSpacing": "initial",
      "lineHeight": "18px",
    },
    "n400": Object {
      "fontFamily": "Helvetica, -apple-system, BlinkMacSystemFont",
      "fontSize": "14px",
      "fontStyle": "normal",
      "fontWeight": 400,
      "letterSpacing": "initial",
      "lineHeight": "18px",
    },
  },
  "s20": Object {
    "n500": Object {
      "fontFamily": "Helvetica, -apple-system, BlinkMacSystemFont",
      "fontSize": "20px",
      "fontStyle": "normal",
      "fontWeight": 500,
      "letterSpacing": "0.29px",
      "lineHeight": "22px",
    },
    "n500_caps": Object {
      "fontFamily": "Helvetica, -apple-system, BlinkMacSystemFont",
      "fontSize": "20px",
      "fontStyle": "normal",
      "fontWeight": 500,
      "letterSpacing": "1px",
      "lineHeight": "1px",
      "textTransform": "uppercase",
    },
    "n500_lower": Object {
      "fontFamily": "Helvetica, -apple-system, BlinkMacSystemFont",
      "fontSize": "20px",
      "fontStyle": "normal",
      "fontWeight": 500,
      "letterSpacing": "initial",
      "lineHeight": "normal",
      "textTransform": "lowercase",
    },
    "nBold": Object {
      "fontFamily": "Helvetica, -apple-system, BlinkMacSystemFont",
      "fontSize": "20rem",
      "fontStyle": "normal",
      "fontWeight": "bold",
      "letterSpacing": "0.29em",
      "lineHeight": "22em",
    },
  },
  "sFooBar": Object {
    "nBold": Object {
      "fontFamily": "Helvetica, -apple-system, BlinkMacSystemFont",
      "fontSize": "FooBar",
      "fontStyle": "normal",
      "fontWeight": "bold",
      "letterSpacing": "0.29em",
      "lineHeight": "22em",
    },
  },
}
`;
